HeatStroke/Services/Collisions/CollisionManager.cpp

```diff
@@ -0,0 +1,210 @@
+//------------------------------------------------------------------------
+// CollisionManager.cpp
+// Author:	David MacIntosh
+//	
+// Manages the interactions between collider components.
+//------------------------------------------------------------------------
+
+#include "CollisionManager.h"
+
+#include "EventManager.h"
+
+using namespace HeatStroke;
+
+CollisionManager* CollisionManager::s_pCollisionManagerInstance = nullptr;
+
+void CollisionManager::CreateInstance()
+{
+	assert(s_pCollisionManagerInstance == nullptr);
+	s_pCollisionManagerInstance = new CollisionManager();
+}
+
+void CollisionManager::DestroyInstance()
+{
+	assert(s_pCollisionManagerInstance != nullptr);
+	delete s_pCollisionManagerInstance;
+	s_pCollisionManagerInstance = nullptr;
+}
+
+CollisionManager* CollisionManager::Instance()
+{
+	assert(s_pCollisionManagerInstance != nullptr);
+	return s_pCollisionManagerInstance;
+}
+
+CollisionManager::CollisionManager()
+	:
+	m_mColliderMap()
+{
+}
+
+CollisionManager::~CollisionManager()
+{
+}
+
+void CollisionManager::RegisterCollider(ComponentCollider* p_pCollider, std::string p_strGuid = "")
+{
+	m_mColliderMap.insert(std::pair<std::string, ComponentCollider*>(p_strGuid, p_pCollider));
+}
+
+void CollisionManager::Update(const float p_fDelta)
+{
+	// Iterate through each pair of colliders
+	for (ColliderMap::iterator i = m_mColliderMap.begin(); i != m_mColliderMap.end(); i++)
+	{
+		for (ColliderMap::iterator j = i; j != m_mColliderMap.end(); j++)
+		{
+			if (i != j)
+			{
+				CheckCollision(i->second, j->second);
+			}
+		}
+	}
+}
+
+void CollisionManager::PreRender()
+{
+}
+
+void CollisionManager::CheckCollision(ComponentCollider* p_pCollider1, ComponentCollider* p_pCollider2)
+{
+	// The structure of this function is actually AIDS but I couldn't think of a better way
```

Comment:
What about creating 3 other function, `CheckSphereSphereCollision`, `CheckSpherePlaneCollision`, and `CheckPlanePlaneCollision`, then have an outer if/else for p_pCollider1 , and within those, an if/else for p_pCollider2.

In psuedocode:
```
if p_pCollider1 is Sphere
  if p_pCollider2 is Sphere
    CheckSphereSphere
  else
    CheckSpherePlane
else
    if p_pCollider2 is Sphere
      CheckSpherePlane
    else
      CheckPlanePlane
```